### PR TITLE
feat(layout): add foundation

### PR DESCRIPTION
**Pull request by Birkbjo**

### CODE REVIEW

File: src/components/loading/index.tsx [hunk 1]
```diff
@@ -0,0 +1,3 @@
+
+
+export { Loader } from './loader'
```
Comment (Mohammer5) [inline src/components/loading/index.tsx:3]:
```suggestion
export { Loader } from './Loader'
```

File: src/pages/overview/AllOverview.tsx [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import React from "react";
+import i18n from "@dhis2/d2-i18n";
+import { GroupOverview } from "../../components/overview";
+
+import { DataElementsCardGroup } from "./DataElements";
```
Comment (Mohammer5) [inline src/pages/overview/AllOverview.tsx:5]:
```suggestion
import { DataElementsCardGroup } from "./dataElements";
```

File: src/pages/overview/index.tsx [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import { RouteObject, Outlet } from "react-router-dom";
+import React, { lazy, Suspense } from "react";
+import { CircularLoader } from "@dhis2/ui";
+import { DefaultLazyLoad } from "../../components";
+
+// React.lazy only works with default exports, so we import directly instead of from index.tsx
+const AllOverview = lazy(() => import("./AllOverview"));
+const DataElements = lazy(() => import("./DataElements"));
```
Comment (Mohammer5) [inline src/pages/overview/index.tsx:8]:
```suggestion
const DataElements = lazy(() => import("./dataElements"));
```

File: src/app/sidebar/sidebar.tsx [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            forceOpen={forceOpen}
+        >
+            {links.map(({ to, label }) => (
+                <SidebarNavLink key={label} to={to} label={label} />
+            ))}
+        </SidenavParent>
+    );
+};
+
+export const Sidebar = ({ links = sidebarLinks }: { links: SidebarLinks }) => {
```
Comment (Mohammer5) [inline src/app/sidebar/sidebar.tsx:69]: I don't think the sidebar needs a `links` prop, we can use the sidebarLinks directly

File: src/app/sidebar/sidebar.tsx [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+    const isFiltered = filterValue !== "";
+    const filteredSidebarLinks = isFiltered
+        ? Object.entries(links).reduce((acc, [key, { label, links }]) => {
+              const filteredLinkItems = links.filter(({ label }) =>
+                  label.toLowerCase().includes(filterValue.toLowerCase())
+              );
+              acc[key] = { label, links: filteredLinkItems };
+              return acc;
+          }, {} as typeof sidebarLinks)
```
Comment (Mohammer5) [inline src/app/sidebar/sidebar.tsx:84]: Creating the filter result is a bit hard to parse and the reduce function is doing multiple things at one. I think it's easier to grasp what's going on with `.map` and `Object.fromEntries` as that's more declarative and less imperative:

```js
Object.fromEntries(
    Object
        .entries(links)
        .map(([key, { label, links }]) => [
            key,
            {
                label,
                links: links.filter(({ label }) =>
                    label.toLowerCase().includes(filterValue.toLowerCase())
                ),
            }
        ])
)
```

File: src/components/loading/Loader.tsx [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import { QueryResponse } from "../../types/query";
+import { NoticeBox } from "@dhis2/ui";
+import i18n from "@dhis2/d2-i18n";
+
+interface LoaderProps {
+    children: React.ReactNode;
+    queryResponse: QueryResponse;
+    label?: string;
+}
+export const Loader = ({ children, queryResponse, label }: LoaderProps) => {
```
Comment (Mohammer5) [inline src/components/loading/Loader.tsx:13]: `<Loader />` needs a `@TODO` comment with a linked Jira issue to get feedback from @cooper-joe so we know how we want to display error and loading states.

Also this component is, right now, only used in the `src/app` component, so it should only be there. `src/components` is - similarly to `src/lib` - for shared components. Until we've gotten feedback from @cooper-joe about how to deal with loading and error states, I'm against making asumptions & adding concretions in shared folders that are not being shared. As soon as we know that this component could/should be used elsewhere, that's when I'd move it to the `components` folder

File: src/pages/overview/index.tsx [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+const OverviewWrapper = () => <Outlet />;
+
+const route: RouteObject = {
```
Comment (Mohammer5) [inline src/pages/overview/index.tsx:12]: Index files' purpose should be to define the public interface of a module (folders in the case of JS). But this file also defines routes. If we decide to keep route definitions in the pages folder (which I'm against), then this should go into its own file.

**Review by Mohammer5 — CHANGES_REQUESTED** [review]
# Centralizing the open state and make nav children stateless

With the simplication of the `forceOpen` and previous state in the `SidenavParent` (see [this comment](https://github.com/dhis2/maintenance-app-beta/pull/323#discussion_r1136409412)), you could centralize the state for which parent is open in the `Sidebar` component by using `useState` that has an array with all opened labels. That way you can remove the entire `HidePreventUnmount` logic / component, which seems unnecessarily complicated to me. The approach of centralizing the state in a single component and than rendering only
stateless components seems to be simpler to me.

# components/utils folder

`src/components/utils` should have a better name. I'd argue for splitting it up into the components that are inside. `src/components` is already an "abstraction" (components that are shared), adding another architectural layer of abstraction makes the architecture more complex & complicated. It's also debatable which component is a `util` and which is not, which introduces ambiguity. With my suggestion to centralize/simplify the state in the sidebar (see one headline above) and subsequently being able to remove `HidePreventUnmount`, we can get rid of this folder as there's nothing in there right now

# Nav filtering

filtering doesn't include group labels (e.g. "Data elements" as search would result in an empty list. I suppose you could argue that "Data element" should be enough, but it wouldn't include the overview pages)

# Components for overview views

With all overview pages being located in `src/pages/overview`, we can move the components inside `src/components/overview` into `src/pages/overview` and co-locate the related code which won't be used anywhere else. Same applies to what can be found in `src/components/card`

# Errors when opening app in the browser

This is happening because some file names are lower case but the import statements use a capitalized folder name and vice versa. It's weird that these errors don't show up on the cli.

```
Compiled with problems:

ERROR in ./src/D2App/App.js 5:104-120

export 'ConfiguredRouter' (imported as 'ConfiguredRouter') was not found in './app/' (possible exports: AppWrapper, Layout)


ERROR in ./src/D2App/app/index.js 3:0-34

export 'Layout' (reexported as 'Layout') was not found in './layout' (module has no exports)


ERROR in ./src/D2App/app/index.js 5:0-44

Module not found: Error: Can't resolve './Router' in '/shared/development/dhis2/maintenance-app-beta/.d2/shell/src/D2App/app'


ERROR in ./src/D2App/app/layout/index.js 3:0-34

Module not found: Error: Can't resolve './Layout' in '/shared/development/dhis2/maintenance-app-beta/.d2/shell/src/D2App/app/layout'
```

File: src/components/loading/Loader.tsx (hunk 6 above)
Comment (Birkbjo) [reply src/components/loading/Loader.tsx:13]: I feel like this is such a generic component that we can assume it will be shared, so I disagree there. I do agree we can talk to Joe though.

File: src/app/sidebar/sidebar.tsx (hunk 4 above)
Comment (Birkbjo) [reply src/app/sidebar/sidebar.tsx:69]: This was mostly done to make it easier to test.

File: src/app/sidebar/sidenav/sidenav.tsx [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+        if (forceOpen != undefined) {
+            beforeForceOpenShowResults.current = showResults;
+            setShowResults(forceOpen);
+        } else {
+            setShowResults(beforeForceOpenShowResults.current);
+        }
+    }, [forceOpen]);
+
+    const isOpen = showResults;
+
```
[the remark this answers is not in the document]
Comment (Birkbjo) [reply src/app/sidebar/sidenav/sidenav.tsx:58]: I actually tried something similar first, but it didn't work out due to some functionality that I think are natural to have. 

The reason for this logic is that `forceOpen` is used to open the the parent when it's filtered, and there's a match. But when the filter is cleared, we want to revert to the previous state, before it was forced open. Which would probably "just" work in your solution if we centralised the open-states in the sidebar. However that brings up my next point:

In your solution the user cannot close a parent while a filter is active, because `forceOpen` would alone decide if it should be open. This is one of the reasons for the complex logic. I guess it could be simplified, but this is what I could come up with that satisfied "my" requirements. Of course these are not strict requirements, but It's what I deemed natural for a feature like this. Maybe it's not worth the "complexity", but in essence it's just keeping a reference to the previous state. So I think this logic would need to stay, regardless if we moved the open-states to the `Sidebar`. 

I will answer the point in the main review thread here, since these are kind of related: 

> With the simplication of the forceOpen and previous state in the SidenavParent (see https://github.com/dhis2/maintenance-app-beta/pull/323#discussion_r1136409412), you could centralize the state for which parent is open in the Sidebar component by using useState that has an array with all opened labels. That way you can remove the entire HidePreventUnmount logic / component, which seems unnecessarily complicated to me. The approach of centralizing the state in a single component and than rendering only
stateless components seems to be simpler to me.

First, I think it's simpler to keep this state in the `SidebarParent`, mostly because the open state depends on the current matched route - and this is more convenient to check there, instead of having to manually `match` for each route. It's also less "wiring", since in that case the `SidebarParent` would need a callback to open the parent, which would cause re-rendering of all items. In this case the performance is negligible, but in general I don't like having a "list" of state unless necessary.
It would certainly be possible, but I don't necessarily think it will be simpler, especially since it doesn't really have any impact on the "complex" logic of this component.

[the remark this answers is not in the document]
Comment (Mohammer5) [reply src/app/sidebar/sidenav/sidenav.tsx:58]: I see. I think this is something we should confirm with @cooper-joe. With the current approach, when you close a parent and then refine your search, that parent will stay closed (I guess that's OK because it's been closed on purpose).

The specs have the following two requirements:

1. Sections are automatically expanded if they contain an item(s) that matches the filter.
2. Sections that contain no items are hidden.

The first point does not mention forcing sections to be opened, just that they should be expanded when there's a filter match (which brings me back to my comment on what should happen to closed sections when changing the search input after having filtered the nav and then closed a section).
The second point mentions that sections without a match should be hidden, which is not the case right now.

Given that sections without items should be hidden, I think I'd

* always keep sections open 
* "force" the user to refine his search if there are too many sections showing (instead of making him close sections)

But as that isn't what specs say either, we need @cooper-joe's input on this topic.

File: src/app/sidebar/sidebar.tsx (hunk 4 above)
Comment (Mohammer5) [reply src/app/sidebar/sidebar.tsx]: That could be done with mocking the sidebarLinks module as well, right? Right now I can't see a test for this though. But I suppose that you could just use the actual list for testing (which on the one hand shouldn't really have an impact on the test code itself and on the other hand actually ensures that tests break when you change the sidebar links when you change them, which enforces that to be a conscious decision)

File: src/components/loading/Loader.tsx (hunk 6 above)
Comment (Mohammer5) [reply src/components/loading/Loader.tsx:12]: Fair enough. I suppose this is also simple enough to be changed very easily if that assumption turns out to be wrong

File: src/app/sidebar/sidenav/sidenav.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (cooper-joe) [reply src/app/sidebar/sidenav/sidenav.tsx:58]: A change in the filtering input constitutes a new attempt at filtering and should "reset" the visible state of a section, in my opinion. That means, whenever a user modifies the filter input, we re-apply the logic of _Sections are automatically expanded if they contain an item(s) that matches the filter_.

File: src/components/loading/Loader.tsx (hunk 6 above)
Comment (cooper-joe) [reply src/components/loading/Loader.tsx:12]: If this is the _first load_ (described [here](https://docs.google.com/document/d/1FJ5NZm1vLCxhB-LXOApKkM5iNDbWkgrYIpSn5lHu97c/edit#heading=h.r5vcxyx16192)) then I think it's OK to assume that it'll be generic. Things get a bit more complex with loading between pages (e.g. loading inside the data-table), but for "load the object list view", we can assume a generic component I think.

File: src/app/sidebar/sidebar.tsx (hunk 5 above)
Comment (Birkbjo) [reply src/app/sidebar/sidebar.tsx]: Did a refactor now, that more resembles something like this.

File: src/app/sidebar/sidebar.tsx (hunk 4 above)
Comment (Mohammer5) [reply src/app/sidebar/sidebar.tsx]: https://dhis2.atlassian.net/browse/DHIS2-15072

File: src/app/router.tsx [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        </Layout>
+    );
+};
+const router = createHashRouter([
+    {
+        // no path means its a layout route
+        element: <LayoutElement />,
+        errorElement: <DefaultErrorRoute />,
+        // TODO: Should we list all routes here, or import children-routes from pages?
+        // PROS: All routes are in one place
```
[the remark this answers is not in the document]
Comment (Mohammer5) [reply src/app/router.tsx:47]: https://dhis2.atlassian.net/browse/DHIS2-15073

File: src/app/sidebar/sidenav/sidenav.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (Mohammer5) [reply src/app/sidebar/sidenav/sidenav.tsx:58]: https://dhis2.atlassian.net/browse/DHIS2-15074

File: d2.config.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
+const config = {
+    type: "app",
+    title: "Maintenance App - Beta",
+    entryPoints: {
+        app: "./src/App.tsx",
```
[the remark this answers is not in the document]
Comment (Mohammer5) [reply d2.config.js:5]: https://dhis2.atlassian.net/browse/DHIS2-15075

[the remark this answers is not in the document]
Comment (Mohammer5) [reply d2.config.js:5]: https://dhis2.atlassian.net/browse/DHIS2-15075

File: src/app/layout/layout.module.css [hunk 11]
```diff
@@ -0,0 +1,34 @@
+@value m-medium from './breakpoints.css';
```
[the remark this answers is not in the document]
Comment (Mohammer5) [reply src/app/layout/layout.module.css:1]: https://dhis2.atlassian.net/browse/DHIS2-15076

File: src/pages/overview/index.tsx (hunk 7 above)
Comment (Mohammer5) [reply src/pages/overview/index.tsx:11]: https://dhis2.atlassian.net/browse/DHIS2-15077

**Review by Mohammer5 — APPROVED** [review]
> Mozafar and I agreed that we want to merge [PR 323](https://github.com/dhis2/maintenance-app-beta/pull/323) asap. So I extracted open comments into separate sub tasks here, so they can be worked on right away after merging the PR.

Immediate follow-up todos: https://dhis2.atlassian.net/browse/DHIS2-15071

**Comment by Birkbjo** [comment]
> src/components/utils should have a better name. I'd argue for splitting it up into the components that are inside. src/components is already an "abstraction" (components that are shared), adding another architectural layer of abstraction makes the architecture more complex & complicated. It's also debatable which component is a util and which is not, which introduces ambiguity. With my suggestion to centralize/simplify the state in the sidebar (see one headline above) and subsequently being able to remove HidePreventUnmount, we can get rid of this folder as there's nothing in there right now

I'm not sure that `utils` and `src/components` would be the same abstraction. It's for `utility components`. I think it's perfectly valid to have a folder like that, and we will most likely end up with components that fit that, instead of creating separate folders for small utility components like these.

We can probably do without `HidePreventUnmount`, but I do think it solves a valid problem and actually simplifies the state handling in this case. But if we don't want to keep the previous state after clearing the filter (as @cooper-joe seems to suggest), it's not needed for now.


## Nav Filtering

> filtering doesn't include group labels (e.g. "Data elements" as search would result in an empty list. I suppose you could argue that "Data element" should be enough, but it wouldn't include the overview pages)

I have added logic for this now. If there's a match in the parent label, all children routes are shown. Do you agree with this logic, @cooper-joe ?

**Comment by Mohammer5** [comment]
> I'm not sure that utils and src/components would be the same abstraction. It's for utility components.

That's not what I was trying to say, but that having two nested abstraction-folders is increasing complexity.

> I think it's perfectly valid to have a folder like that, and we will most likely end up with components that fit that, instead of creating separate folders for small utility components like these.

Another point I made is that it can be debatable which component is a util and which is not, which introduces ambiguity. I think `HidePreventUnmount` is a very good example for what I'm trying to point out. As a developer, I know that `src/components` will contain shared components, the abstraction level is quite clear. Then inside that folder, there's the utils folder which I wouldn't know what to find inside conceptually. Why would `HidePreventUnmount` be a utility? What's the difference between that component and `Loader` for example, which could be considered a utility component but is not in the `utils` folder. Just like `DefaultLazyLoad`.

Considering my comment to move what's inside `components/card` & `components/overview` into `pages/overview` (as that is the only place where these will be used, they're not really that "shared"), everything that'd be left in `components` could be considered a utility component.

So having `utils` seems very ambiguous to me and therefore also reduce discoverability due to the extra level of nesting. I can't see the advantage or necessity of that folder yet, but on the other hand I can see disadvantages and lack of clarity.

> We can probably do without HidePreventUnmount, but I do think it solves a valid problem and actually simplifies the state handling in this case. But if we don't want to keep the previous state after clearing the filter (as @cooper-joe seems to suggest), it's not needed for now.

I can't see the component being used anywhere right now.

**Comment by Birkbjo** [comment]
> Another point I made is that it can be debatable which component is a util and which is not, which introduces ambiguity.

> Why would HidePreventUnmount be a utility? What's the difference between that component and Loader for example, which could be considered a utility component but is not in the utils folder. Just like DefaultLazyLoad.

Yeah I agree that it can be a bit arbitrary. I'm not sure if it's ambiguous though - it's just a decision where it lives. And everything we do are decisions right? And can look arbitrary if you don't know the context or reason behind a decision. But that doesn't mean it's ambiguous. `utils` is just a catch-all for small utility components that doesn't "deserve" it's own folder. I do think both `Loader` and `DefaultLazyLoad` could be in a `utils`-folder, but as they have styles attached, and thus not in one file I decided to move it to their own folders. I think introducing a folder for just one file is a bit overkill, and I wouldn't even be sure what to call the concept of `HidePreventUnmount` - it's just a small utility component.

**Comment by Mohammer5** [comment]
@Birkbjo I've created a separate issue (#324 / [DHIS2-15080](https://dhis2.atlassian.net/browse/DHIS2-15080)) so we can talk about it. I'll merge this PR to unblock work.
